### PR TITLE
[Python] Optional Pandas Date as datetime

### LINKED ISSUE #4428: Control over Pandas type conversion
I was wondering if there is any way to control the type conversion with the Python client when converting results to a Pandas DataFrame. Specifically with columns of the [date](https://duckdb.org/docs/sql/data_types/date) type it looks like by default these are converted to Pandas `datetime64[ns]` when I would prefer a Pandas object column containing [Python `datetime.date`](https://docs.python.org/3/library/datetime.html#date-objects) values. This is is actually the default behavior if you do `conn.execute(...).arrow().to_pandas()` but I would like to avoid the intermediate Arrow conversion. Pyarrow exposes control over this with a [`date_as_object` flag](https://arrow.apache.org/docs/python/pandas.html#date-types). It would be nice if there were something similar for direct Pandas conversion.

### PR DESCRIPTION
**Pull request by pdet**
Fix: #4428

### CODE REVIEW

File: tools/pythonpkg/tests/fast/pandas/test_date_as_datetime.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import pandas as pd
+import duckdb
+import datetime
+
+def run_checks(df):
+    assert type(df['d'][0]) is datetime.date
```
Comment (Mause) [inline tools/pythonpkg/tests/fast/pandas/test_date_as_datetime.py:6]:
```suggestion
    assert isinstance(df['d'][0], datetime.date)
```

File: tools/pythonpkg/src/pyrelation.cpp [hunk 2]
```diff
@@ -129,8 +129,10 @@ void DuckDBPyRelation::Initialize(py::handle &m) {
 	    .def("fetchall", &DuckDBPyRelation::Fetchall, "Execute and fetch all rows as a list of tuples")
 	    .def("fetchnumpy", &DuckDBPyRelation::FetchNumpy,
 	         "Execute and fetch all rows as a Python dict mapping each column to one numpy arrays")
-	    .def("df", &DuckDBPyRelation::ToDF, "Execute and fetch all rows as a pandas DataFrame")
-	    .def("to_df", &DuckDBPyRelation::ToDF, "Execute and fetch all rows as a pandas DataFrame")
+	    .def("df", &DuckDBPyRelation::ToDF, "Execute and fetch all rows as a pandas DataFrame",
+	         py::arg("date_as_datetime") = false)
```
Comment (Mause) [inline tools/pythonpkg/src/pyrelation.cpp:133]: Maybe we can amend the docstrings to document the parameter? Preferably using the Sphinx RST syntax so the docs look nice

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/duckdb-stubs/__init__.pyi [hunk 3]
```diff
@@ -133,7 +133,7 @@ class DuckDBPyRelation:
     def std(self, std_aggr: str, group_expr: str = ...) -> DuckDBPyRelation: ...
     def sum(self, sum_aggr: str, group_expr: str = ...) -> DuckDBPyRelation: ...
     def to_arrow_table(self, batch_size: int = ...) -> pyarrow.lib.Table: ...
-    def to_df(self) -> pandas.DataFrame: ...
+    def to_df(self, date_as_datetime: bool) -> pandas.DataFrame: ...
```
Comment (Mause) [inline tools/pythonpkg/duckdb-stubs/__init__.pyi:136]: Can you check the default is correctly declared? It should be populated in the stubs pyi as well

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/duckdb-stubs/__init__.pyi [hunk 4]
```diff
@@ -161,13 +161,13 @@ class DuckDBPyResult:
     def arrow(self, chunk_size: int = ...) -> pyarrow.lib.Table: ...
     def close(self) -> None: ...
     def description(self) -> list: ...
-    def df(self) -> pandas.DataFrame: ...
+    def df(self, date_as_datetime: bool) -> pandas.DataFrame: ...
     def fetch_arrow_reader(self, approx_batch_size: int) -> pyarrow.lib.RecordBatchReader: ...
     def fetch_arrow_table(self, chunk_size: int = ...) -> pyarrow.lib.Table: ...
-    def fetch_df(self) -> pandas.DataFrame: ...
-    def fetch_df_chunk(self, num_of_vectors: int = ...) -> pandas.DataFrame: ...
+    def fetch_df(self, date_as_datetime: bool) -> pandas.DataFrame: ...
```
Comment (Mytherin) [inline tools/pythonpkg/duckdb-stubs/__init__.pyi:167]: Boolean parameters are in general a bit dangerous - how do other systems (e.g. arrow) do this? Perhaps having a global setting is more readable/clear?

Comment (pdet) [reply tools/pythonpkg/duckdb-stubs/__init__.pyi:167]: Arrow has a boolean as well (https://arrow.apache.org/docs/python/pandas.html#date-types)

I also think global is cleaner, but on the other hand, is good to follow stds. If you strongly prefer global I'm fine going with it.

Comment (Mytherin) [reply tools/pythonpkg/duckdb-stubs/__init__.pyi:167]: Can we somehow enforce this to be a named parameter, so you would e.g. have to type `fetch_df(date_as_datetime=True)`?

Should we also use the same naming (i.e. `date_as_object`)?

Comment (Mause) [reply tools/pythonpkg/duckdb-stubs/__init__.pyi:167]: You can yeah, like this:

```py
 def fetch_df(self, *, date_as_datetime: bool)
```
Not sure if pybind11 supports that tho

Comment (pdet) [reply tools/pythonpkg/duckdb-stubs/__init__.pyi:167]: I think you had a good point before as well, it seems to me a user would either emit all dfs one way or the other, so maybe a global config is cleaner than adding this type-specific option to ```df()```.
If no one is strongly opposed to that, I think I'll go that way.

**Comment by pdet** [comment]
@Mause and @dylanscott what you guys think is preferable, date_as_object parameter or global config?

I'm fine with either.

**Comment by pdet** [comment]
@Mytherin can you restart these jobs?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**